src/bvals/comms/coalesced_buffers.hpp

```diff
@@ -0,0 +1,207 @@
+                      const std::shared_ptr<Variable<Real>> &var);
+
+  void AllocateCoalescedBuffer();
+
+  bool IsAvailableForWrite() {
+    return sparse_status_buffer.IsAvailableForWrite() &&
+           coalesced_comm_buffer.IsAvailableForWrite();
+  }
+
+  ParArray1D<BndId> &GetBndIdsOnDevice(const std::set<Uid_t> &vars,
+                                       int *pcomb_size = nullptr);
+
+  void PackAndSend(const std::set<Uid_t> &vars);
+
+  bool TryReceiveAndUnpack(const std::set<Uid_t> &vars);
+};
+
+struct CoalescedBuffersRank {
+  using coalesced_message_structure_t = std::vector<BndId>;
+  using buf_t = BufArray1D<Real>;
+
+  // Rank that these buffers communicate with
+  BoundaryType b_type;
+  int other_rank;
+
+  // map from partion id to coalesced message structure for communication
+  // partition id of the sender will be the mpi tag we use
+  bool buffers_built{false};
+
+  std::map<int, CoalescedBuffer> coalesced_bufs;
+
+  static constexpr int nglobal{1};
+  static constexpr int nper_part{3};
+
+  using com_buf_t = CommBuffer<std::vector<int>>;
+  com_buf_t message;
+
+  mpi_comm_t comm_;
+  Mesh *pmesh;
+  bool sender{true};
+
+  explicit CoalescedBuffersRank(int o_rank, BoundaryType b_type, bool send,
+                                mpi_comm_t comm, Mesh *pmesh);
+
+  void AddSendBuffer(int partition, MeshBlock *pmb, const NeighborBlock &nb,
+                     const std::shared_ptr<Variable<Real>> &var);
+
+  bool TryReceiveBufInfo();
+
+  void ResolveAndSendBufInfo();
+
+  void PackAndSend(MeshData<Real> *pmd);
+
+  bool TryReceiveAndUnpack(MeshData<Real> *pmd, int partition);
+
+  bool IsAvailableForWrite(MeshData<Real> *pmd);
+};
+
+struct CoalescedComms {
+  // Combined buffers for each rank
+  std::map<std::pair<int, BoundaryType>, CoalescedBuffersRank> coalesced_send_buffers;
+  std::map<std::pair<int, BoundaryType>, CoalescedBuffersRank> coalesced_recv_buffers;
+
+  std::map<BoundaryType, mpi_comm_t> comms_;
+
+  Mesh *pmesh;
+
+  explicit CoalescedComms(Mesh *pmesh) : pmesh(pmesh) {
+    // TODO(LFR): Switch to a different communicator for each BoundaryType pair
+    for (auto b_type :
+         {BoundaryType::any, BoundaryType::flxcor_send, BoundaryType::gmg_same,
+          BoundaryType::gmg_restrict_send, BoundaryType::gmg_prolongate_send}) {
+      auto &comm = comms_[b_type];
+#ifdef MPI_PARALLEL
+      PARTHENON_MPI_CHECK(MPI_Comm_dup(MPI_COMM_WORLD, &comm));
+#else
+      comm = 0;
+#endif
+    }
+  }
+
+  ~CoalescedComms() {
+#ifdef MPI_PARALLEL
+    for (auto &[b_type, comm] : comms_)
+      PARTHENON_MPI_CHECK(MPI_Comm_free(&comm));
+#endif
+  }
+
+  void clear() {
+    bool can_delete;
+    do {
+      can_delete = true;
+      for (auto &[p, cbr] : coalesced_send_buffers) {
+        for (auto &[r, cbrp] : cbr.coalesced_bufs) {
+          can_delete = cbrp.IsAvailableForWrite() && can_delete;
+        }
+      }
+    } while (!can_delete);
```

Comment:
What's going on here?

Reply:
You need to check that communication on all of the send buffers is completed before you delete them.

Reply:
Conceptually I understand but I don't understand the do while loop. Is the idea you just wait for all of them? Could this be a please where the code could hang? It might be worth having some max number of iters you're allowed to wait and then failing or printing an error message or something.

Reply:
yeah, you just need to wait until MPI is done with the buffers. As you say, this is a possible hang location, so I added a max number of iterations.